src/main/kotlin/org/rust/lang/core/psi/ext/RsElement.kt

```diff
@@ -138,3 +139,26 @@ fun RsElement.findInScope(name: String, ns: Set<Namespace>): PsiElement? {
 
 fun RsElement.hasInScope(name: String, ns: Set<Namespace>): Boolean =
     findInScope(name, ns) != null
+
+
+fun <T> RsElement.createCachedResult(value: T): CachedValueProvider.Result<T> {
+    val structureModificationTracker = project.rustStructureModificationTracker
+
+    return when {
+        // The case of injected language. Injected PSI don't have it's own event system, so can only
+        // handle evens from outer PSI. For example, Rust language is injected to Kotlin's string
+        // literal. If a user change the literal, we can only be notified that the literal is changed.
+        // So we have to invalidate the cached value on any PSI change
+        containingFile.virtualFile is VirtualFileWindow -> CachedValueProvider.Result.create(value, PsiModificationTracker.MODIFICATION_COUNT)
+
+        // Invalidate cached value of code fragment on any PSI change
+        this is RsCodeFragment -> CachedValueProvider.Result.create(value, PsiModificationTracker.MODIFICATION_COUNT)
+
+        // CachedValueProvider.Result can accept a ModificationTracker as a dependency, so the
+        // cached value will be invalidated if the modification counter is incremented.
+        else -> {
+            val modificationTracker = contextOrSelf<RsModificationTrackerOwner>()?.modificationTracker
```

Comment:
> Currently, the inspection does not seem to run after every code change, I'm not sure what should be changed in the caching implementation, any hints?

I haven't read the whole implementation yet but it's likely the reason of the described problem.
`RsModificationTrackerOwner` and `rustStructureModificationTracker` are supposed to be used for computation that don't depend on code inside function body. But "Unused import" functionality depends on it that leads to issue with incorrect updating. Just use `PsiModificationTracker.MODIFICATION_COUNT`

Reply:
Thank you :)